### PR TITLE
Count terminating pods when deleting active pods for failed jobs

### LINKED ISSUE #125089: Job controller reports the count of terminating pods with unnecessary delay
### What happened?

When Job controller is terminating and deletes all pods the counter in the `status.terminating` field does not reflect this properly. 

### What did you expect to happen?

The counter for the terminating pods is updated as soon as Job controller has this information. Similarly as for the counter of active pods.

### How can we reproduce it (as minimally and precisely as possible)?

0. watch the job & pod updates with the commands:
`kubectl get jobs -w --output-watch-events -ocustom-columns=EVENT:.type,NAME:.object.metadata.name,STATUS:.object.status | ts "%Y-%m-%d %H:%M:%.S"`
`kubectl get pods -w --output-watch-events -ocustom-columns=EVENT:.type,NAME:.object.metadata.name,PHASE:.object.status.phase,FINALIZERS:.object.metadata.finalizers,DELETION:object.metadata.deletionTimestamp,CONTAINERS:.object.status.containerStatuses | ts "%Y-%m-%d %H:%M:%.S"`
1. Run the Job:
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: indexed-job
  labels:
    jobgroup: indexedjob
spec:
  completions: 2
  parallelism: 2
  backoffLimit: 0
  completionMode: Indexed
  podReplacementPolicy: Failed
  template:
    spec:
      restartPolicy: Never
      containers:
      - name: 'worker'
        image: python
        readinessProbe:
          httpGet:
            path: /non-existent-endpoint
            port: 80
          initialDelaySeconds: 1
          periodSeconds: 10
          failureThreshold: 10
        command:
        - python3
        - -c
        - |
          import os, sys, time
          print("Hello world")
          if int(os.environ.get("JOB_COMPLETION_INDEX")) == 0:
            time.sleep(60)
            sys.exit(1)
          time.sleep(300)
          sys.exit(0)
```
3. When the job transitions you can observe updates like this:
For Job:
```
2024-05-23 13:52:11.053575 EVENT   NAME          STATUS
2024-05-23 13:52:11.053833 ADDED   indexed-job   map[]
2024-05-23 13:52:11.076666 MODIFIED   indexed-job   map[active:2 ready:0 startTime:2024-05-23T11:52:11Z terminating:0 uncountedTerminatedPods:map[]]
2024-05-23 13:53:15.315997 MODIFIED   indexed-job   map[ready:0 startTime:2024-05-23T11:52:11Z terminating:0 uncountedTerminatedPods:map[failed:[b237e4f0-dc57-4349-8aa6-7bc381266aed d4dfe535-4b6d-4374-8e9d-b73c9823cd35]]]
2024-05-23 13:53:15.328668 MODIFIED   indexed-job   map[conditions:[map[lastProbeTime:2024-05-23T11:53:15Z lastTransitionTime:2024-05-23T11:53:15Z message:Job has reached the specified backoff limit reason:BackoffLimitExceeded status:True type:Failed]] failed:2 ready:0 startTime:2024-05-23T11:52:11Z terminating:0 uncountedTerminatedPods:map[]]
```
**Issue**: the job reports `terminating: 0` even though it already deleted the second pod, because it is enumerated in the 
`uncountedTerminatedPods`. Also, it is inconsistent, `active` is reported as `0` and `terminating` is 0 as well, while setting `Failed` condition, the pods are in fact still terminating. This can be visible from the pod updates:
```
2024-05-23 13:52:11.063000 EVENT   NAME                  PHASE     FINALIZERS                           DELETION   CONTAINERS
2024-05-23 13:52:11.063188 ADDED   indexed-job-0-wz49q   Pending   [batch.kubernetes.io/job-tracking]   <none>     <none>
2024-05-23 13:52:11.066376 MODIFIED   indexed-job-0-wz49q   Pending   [batch.kubernetes.io/job-tracking]   <none>     <none>
2024-05-23 13:52:11.069335 ADDED      indexed-job-1-jhl9t   Pending   [batch.kubernetes.io/job-tracking]   <none>     <none>
2024-05-23 13:52:11.076759 MODIFIED   indexed-job-1-jhl9t   Pending   [batch.kubernetes.io/job-tracking]   <none>     <none>
2024-05-23 13:52:11.079168 MODIFIED   indexed-job-0-wz49q   Pending   [batch.kubernetes.io/job-tracking]   <none>     [map[image:python imageID: lastState:map[] name:worker ready:false restartCount:0 started:false state:map[waiting:map[reason:ContainerCreating]]]]
2024-05-23 13:52:11.089661 MODIFIED   indexed-job-1-jhl9t   Pending   [batch.kubernetes.io/job-tracking]   <none>     [map[image:python imageID: lastState:map[] name:worker ready:false restartCount:0 started:false state:map[waiting:map[reason:ContainerCreating]]]]
2024-05-23 13:52:13.031692 MODIFIED   indexed-job-0-wz49q   Running   [batch.kubernetes.io/job-tracking]   <none>     [map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:true state:map[running:map[startedAt:2024-05-23T11:52:12Z]]]]
2024-05-23 13:52:14.034589 MODIFIED   indexed-job-1-jhl9t   Running   [batch.kubernetes.io/job-tracking]   <none>     [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:true state:map[running:map[startedAt:2024-05-23T11:52:13Z]]]]
2024-05-23 13:53:13.140294 MODIFIED   indexed-job-0-wz49q   Running   [batch.kubernetes.io/job-tracking]   <none>     [map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f exitCode:1 finishedAt:2024-05-23T11:53:12Z reason:Error startedAt:2024-05-23T11:52:12Z]]]]
2024-05-23 13:53:14.300875 MODIFIED   indexed-job-0-wz49q   Failed    [batch.kubernetes.io/job-tracking]   <none>     [map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f exitCode:1 finishedAt:2024-05-23T11:53:12Z reason:Error startedAt:2024-05-23T11:52:12Z]]]]
2024-05-23 13:53:15.142013 MODIFIED   indexed-job-0-wz49q   Failed    [batch.kubernetes.io/job-tracking]   <none>     [map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f exitCode:1 finishedAt:2024-05-23T11:53:12Z reason:Error startedAt:2024-05-23T11:52:12Z]]]]
2024-05-23 13:53:15.308553 MODIFIED   indexed-job-1-jhl9t   Running   [batch.kubernetes.io/job-tracking]   2024-05-23T11:53:45Z   [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:true state:map[running:map[startedAt:2024-05-23T11:52:13Z]]]]
2024-05-23 13:53:15.324258 MODIFIED   indexed-job-0-wz49q   Failed    <none>                               <none>                 [map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://d4b0a03a1aef94a10d4397a0d82a79cc6edbf32293de038efc17d16e87d9e75f exitCode:1 finishedAt:2024-05-23T11:53:12Z reason:Error startedAt:2024-05-23T11:52:12Z]]]]
2024-05-23 13:53:15.324497 MODIFIED   indexed-job-1-jhl9t   Running   <none>                               2024-05-23T11:53:45Z   [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:true state:map[running:map[startedAt:2024-05-23T11:52:13Z]]]]
2024-05-23 13:53:45.569196 MODIFIED   indexed-job-1-jhl9t   Failed    <none>                               2024-05-23T11:53:45Z   [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 exitCode:137 finishedAt:2024-05-23T11:53:45Z reason:Error startedAt:2024-05-23T11:52:13Z]]]]
2024-05-23 13:53:46.198066 MODIFIED   indexed-job-1-jhl9t   Failed    <none>                               2024-05-23T11:53:45Z   [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 exitCode:137 finishedAt:2024-05-23T11:53:45Z reason:Error startedAt:2024-05-23T11:52:13Z]]]]
2024-05-23 13:53:46.205463 MODIFIED   indexed-job-1-jhl9t   Failed    <none>                               2024-05-23T11:53:15Z   [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 exitCode:137 finishedAt:2024-05-23T11:53:45Z reason:Error startedAt:2024-05-23T11:52:13Z]]]]
2024-05-23 13:53:46.207674 DELETED    indexed-job-1-jhl9t   Failed    <none>                               2024-05-23T11:53:15Z   [map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 image:docker.io/library/python:latest imageID:docker.io/library/python@sha256:3966b81808d864099f802080d897cef36c01550472ab3955fdd716d1c665acd6 lastState:map[] name:worker ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://e1f78aec0a1b9b1735f77192f2878afec72aecabb8afc7a40e065f425380d800 exitCode:137 finishedAt:2024-05-23T11:53:45Z reason:Error startedAt:2024-05-23T11:52:13Z]]]]
```
Note that the second pod gets the deletionTimestamp at `2024-05-23 13:53:15.324497` which is prior to the Job controller setting the `Failed` condition.

I think when fix we should have the `terminating>0` for the last status updates for the Job controller above.

### Anything else we need to know?

I think the proper solution is to increment the counter of `terminating` pods [here](https://github.com/kubernetes/kubernetes/blob/c9a1a0a3b8acb70cba30d6f4ea59505b4564b4d9/pkg/controller/job/job_controller.go#L895-L902), where we decrement the `active` pods.

Also, this fix is related to https://github.com/kubernetes/kubernetes/issues/123775. If we have the fix we could delay setting the `Failed` condition until `terminating=0`. Otherwise we may observe `terminating=0` because the update of the counter is delayed.

### Kubernetes version

<details>
Server Version: v1.30.0
</details>

### Cloud provider

<details>
Kind
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by dejanzele**
#### What type of PR is this?

/kind bug

/sig apps

#### What this PR does / why we need it:

Count terminating pods when deleting active pods for a failed job.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/125089

#### Special notes for your reviewer:

This is prerequisite for https://github.com/kubernetes/kubernetes/issues/123775

#### Does this PR introduce a user-facing change?

```release-note
.status.terminating field now gets tracked faster when active Pods are deleted, specifically when Job is failed, gets suspended or has too many active pods
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated
```

### CODE REVIEW

**Review by mimowo — COMMENTED** [review]
LGTM as the fix for the main scenario of terminating job. However, I think for consistency we should also include the update of terminating when [suspending](https://github.com/kubernetes/kubernetes/blob/7ba244f50e57257e9d435b42f3b0f17a92a0611a/pkg/controller/job/job_controller.go#L1501-L1508), and [deleting excess pods](https://github.com/kubernetes/kubernetes/blob/7ba244f50e57257e9d435b42f3b0f17a92a0611a/pkg/controller/job/job_controller.go#L1551-L1561).

**Comment by mimowo** [comment]
Please also add reference to the [Pod replacement policy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3939-allow-replacement-when-fully-terminated)

File: pkg/controller/job/job_controller_test.go [hunk 1]
```diff
@@ -3640,7 +3640,7 @@ func TestSyncJobWithJobSuccessPolicy(t *testing.T) {
 			wantStatus: batch.JobStatus{
 				Failed:                  1,
 				Succeeded:               1,
-				Terminating:             ptr.To[int32](0),
+				Terminating:             ptr.To[int32](2),
```
Comment (mimowo) [inline pkg/controller/job/job_controller_test.go:3643]: This is a couple of changes, maybe better introduce `enableJobPodReplacementPolicy` as for other features.

Also, I would suggest to also add 2 test cases to "TestControllerSyncJob" - one with the `JobPodReplacementPolicy` feature gate disabled, and one with enabled?

**Comment by mimowo** [comment]
> @mimowo my idea was to do the fix for suspend in a followup PR to minimize the change in this PR. I can do it in the same PR if you'd like.

I see, I think the same PR is fine given the change is still relatively small, all places to all three changes are analogous (probably), so it might be beneficial to the reader to see them all applied.

**Comment by dejanzele** [comment]
I agree, I'll add the additional changes in this PR

**Review by mimowo — COMMENTED** [review]
@dejanzele thanks for updating the PR, are all logic changes covered by at least one change in tests?

**Comment by dejanzele** [comment]
@mimowo I will now address your comment https://github.com/kubernetes/kubernetes/pull/125175#discussion_r1618448563, and it should be ready for final review

**Comment by dejanzele** [comment]
@mimowo and reviewers

In `TestControllerSyncJob` now we verify the following:
1. suspending a job
- `suspending a job with satisfied expectations` - default PodReplacementPolicy feature enabled is `false`, `terminating` should be `nil`
- `suspending a job with satisfied expectations; PodReplacementPolicy enabled` - `terminating` should be `2`
- `suspending a job with unsatisfied expectations` - default PodReplacementPolicy feature enabled is `false`, terminating should be `nil`
- `suspending a job with unsatisfied expectations; PodReplacementPolicy enabled` - `terminating` should be `0`
2. too many active pods
- `too many active pods` - default PodReplacementPolicy is `false`, `terminating` should be `nil`
- `too many active pods; PodReplacementPolicy enabled` - `terminating` should be `1`
3. job failed
- `indexed job failed` - default PodReplacementPolicy feature state is `false`, `terminating` should be `nil` (this is the actual bug)
- `count terminating pods when indexed job fails and PodReplacementPolicy enabled` - `terminating` should be `1` (this verifies the fix guarded by the PodReplacementPolicy Feature Gate)

In `TestSyncJobWithJobSuccessPolicy` I also added test cases how enabling|disabling PodReplacementPolicy impacts `JobSuccessPolicy` and `JobFailurePolicy` Feature Gates

**Review by mimowo — COMMENTED** [review]
/lgtm
Thanks! 
/assign @atiratree @alculquicondor

**Review by alculquicondor — COMMENTED** [review]
Do I understand correctly that you are only fixing temporary transitions?

Because, IIUC, even if we don't increment the counters as this PR does, the counters would be incremented by the next syncJob, as Pods will start terminating.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Comment by mimowo** [comment]
> Do I understand correctly that you are only fixing temporary transitions?

Yes, correct. However, computing the `terminating`  without lag seems like an important pre-requisite for https://github.com/kubernetes/kubernetes/issues/123775, if we want to delay setting `Failed` until `terminating=0`. Otherwise we may get a false impression that `terminating=0`, but it will be just due to the lag.

> Because, IIUC, even if we don't increment the counters as this PR does, the counters would be incremented by the next syncJob, as Pods will start terminating.

Correct.

**Comment by dejanzele** [comment]
Release notes updated per @alculquicondor comment https://github.com/kubernetes/kubernetes/pull/125546#issuecomment-2181713282

**Pull request merged**